### PR TITLE
VTGate Buffering: Use a more accurate heuristic for determining if we're doing a reshard

### PR DESCRIPTION
**Pull request by mattlord**
## Description

There are two primary cases today where [vtgate performs query buffering](https://vitess.io/docs/17.0/reference/features/vtgate-buffering/):
 1. The primary tablet for a shard is not serving — meaning that we're probably in the middle of a reparent operation of some kind
 2. We detect that there's a resharding operation happening 

The problem was that:
1. We checked for a resharding event type before a primary tablet not serving event type: https://github.com/vitessio/vitess/blob/1fb23351fa698e6b0113f3c25e3b0ff3f6e41a6e/go/vt/vtgate/tabletgateway.go#L286-L295
2. The heuristic used for determining if a reshard was happening was not accurate: https://github.com/vitessio/vitess/blob/1fb23351fa698e6b0113f3c25e3b0ff3f6e41a6e/go/vt/discovery/keyspace_events.go#L151-L159

This PR adjusts the heuristic to more accurately look specifically for the case where the keyspace has overlapping shards, which should only be the case when there's an active [`Reshard`](https://vitess.io/docs/17.0/reference/vreplication/reshard/) in progress.

The new unit test fails on main:
```
❯ go test -timeout 30s -run ^TestKeyspaceEventTypes$ vitess.io/vitess/go/vt/discovery
ERROR: logging before flag.Parse: W0827 16:52:23.110207   97333 log.go:39] Failed to read in config : Config File "vtconfig" Not Found in "[/Users/matt/git/vitess/go/vt/discovery]". This is optional, and can be ignored if you are not using config files. For a detailed explanation, see https://github.com/vitessio/vitess/blob/main/doc/viper/viper.md#config-files.
--- FAIL: TestKeyspaceEventTypes (0.00s)
    --- FAIL: TestKeyspaceEventTypes/sharded_primary_not_serving (0.00s)
        keyspace_events_test.go:265:
            	Error Trace:	/Users/matt/git/vitess/go/vt/discovery/keyspace_events_test.go:265
            	Error:      	Not equal:
            	            	expected: true
            	            	actual  : false
            	Test:       	TestKeyspaceEventTypes/sharded_primary_not_serving
            	Messages:   	TargetIsBeingResharded should return false
FAIL
FAIL	vitess.io/vitess/go/vt/discovery	0.568s
FAIL
```

>**Note**
> The [manual test used here](https://gist.github.com/mattlord/ea49abc347b02b2ee22db4b776800a8d) has the expected results in this PR branch. Meaning that for the `commerce/-80` shard buffering is engaged, we buffer 1 query, that buffer is quickly drained when that shard's primary tablet is restarted because of NewPrimarySeen rather than after 20s because of MaxDurationExceeded:
```
Post restart buffering stats:
"BufferFailoverDurationSumMs": {"commerce.-80": 547, "commerce.80-": 20001},
"BufferLastFailoverDurationMs": {"commerce.-80": 547, "commerce.80-": 20001},
"BufferLastRequestsDryRunMax": {"commerce.-80": 0, "commerce.80-": 0},
"BufferLastRequestsInFlightMax": {"commerce.-80": 1, "commerce.80-": 1},
"BufferRequestsBuffered": {"commerce.-80": 1, "commerce.80-": 2},
"BufferRequestsBufferedDryRun": {"commerce.80-": 0, "commerce.-80": 0},
"BufferRequestsDrained": {"commerce.-80": 1, "commerce.80-": 1},
"BufferRequestsEvicted": {"commerce.-80.ContextDone": 0, "commerce.-80.BufferFull": 0, "commerce.-80.WindowExceeded": 0, "commerce.80-.ContextDone": 0, "commerce.80-.BufferFull": 0, "commerce.80-.WindowExceeded": 1},
"BufferRequestsSkipped": {"commerce.-80.BufferFull": 0, "commerce.-80.Disabled": 0, "commerce.-80.Shutdown": 0, "commerce.-80.LastReparentTooRecent": 0, "commerce.-80.LastFailoverTooRecent": 0, "commerce.80-.Disabled": 0, "commerce.80-.Shutdown": 0, "commerce.80-.BufferFull": 0, "commerce.80-.LastReparentTooRecent": 0, "commerce.80-.LastFailoverTooRecent": 1168},
"BufferSize": 1000,
"BufferStarts": {"commerce.-80": 1, "commerce.80-": 1},
"BufferStops": {"commerce.80-.NewPrimarySeen": 0, "commerce.80-.MaxDurationExceeded": 1, "commerce.80-.Shutdown": 0, "commerce.-80.ReshardingComplete": 0, "commerce.-80.NewPrimarySeen": 1, "commerce.-80.MaxDurationExceeded": 0, "commerce.-80.Shutdown": 0, "commerce.80-.ReshardingComplete": 0},
"BufferUtilizationDryRunSum": {"commerce.-80": 0, "commerce.80-": 0},
"BufferUtilizationSum": {"commerce.-80": 0, "commerce.80-": 0},
```

## Related Issue(s)

  - Fixes: https://github.com/vitessio/vitess/issues/13854

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation is not required

### CODE REVIEW

File: examples/common/scripts/vtgate-up.sh [hunk 1]
```diff
@@ -39,6 +39,7 @@ vtgate \
   --tablet_types_to_wait PRIMARY,REPLICA \
   --service_map 'grpc-vtgateservice' \
   --pid_file $VTDATAROOT/tmp/vtgate.pid \
+  --enable_buffer \
```
Comment (mattlord) [inline examples/common/scripts/vtgate-up.sh:42]: This merely makes it easier for users to test buffering related behavior in the local examples moving forward.

**Review by vmg — APPROVED** [review]
I love the new heuristic of overlapping keyspace ranges. Really seems like the best way to detect a re-shard in process. Great job!

File: go/vt/discovery/keyspace_events_test.go [hunk 2]
```diff
@@ -55,6 +58,220 @@ func TestSrvKeyspaceWithNilNewKeyspace(t *testing.T) {
...
+// And we should consider the cause to be a primary not serving when the
+// following conditions exist:
+// 1. The keyspace is inconsistent (in the middle of an availability event)
+// 2. The target tablet is a primary
+// 3. The target tablet is not serving
+// 4. The shard's externallyReparented time is not 0
+// 5. The shard's currentPrimary state is not nil
+// We should never consider both as a possible cause given the same
+// keyspace state.
+func TestKeyspaceEventTypes(t *testing.T) {
```
Comment (rohit-nayak-ps) [inline go/vt/discovery/keyspace_events_test.go:78]: ❤️

Comment (deepthi) [reply go/vt/discovery/keyspace_events_test.go:78]: Nice test description 💯

**Review by deepthi — COMMENTED** [review]
@mattlord given
> We checked for a resharding event type before a primary tablet not serving event type

Is it possible to rearrange the code to check for the "primary tablet not serving event type" first?

**Comment by mattlord** [comment]
> @mattlord given
> 
> > We checked for a resharding event type before a primary tablet not serving event type
> 
> Is it possible to rearrange the code to check for the "primary tablet not serving event type" first?

Yeah, we could. I thought about that too. It's probably a nice optimization given that it's a simpler check and I would think that a primary being down is much more common.

If you like the idea, I can create a quick follow-up PR for that.

**Pull request merged**